### DISCUSSION TITLE
Nested OneToOne insert is not taking owner's primary key

### DISCUSSION THREAD

**Discussion by kyllerss** [mikro-orm/mikro-orm#3570] [Q&A] [answered]
I am new to Mikro ORM. So I am clearly missing something fundamental here. 
I am trying to save a `OneToOne` relation, but am I am erroring-out because the non-owning part of the relation isn't getting it's primary key set. 

Here is the schema in question:

```
CREATE TABLE public.track_metrics (
    track_uuid uuid NOT NULL,   // <-- This points to the 'owning' entity
    ...
);
ALTER TABLE ONLY public.track_metrics
    ADD CONSTRAINT track_metrics_pk PRIMARY KEY (track_uuid);
ALTER TABLE ONLY public.track_metrics
    ADD CONSTRAINT track_metrics_fk FOREIGN KEY (track_uuid) REFERENCES public.tracks(uuid);

CREATE TABLE public.tracks (
    uuid uuid DEFAULT gen_random_uuid() NOT NULL,
    ...
);
ALTER TABLE ONLY public.tracks
    ADD CONSTRAINT tracks_pkey PRIMARY KEY (uuid);

```

Here is the definition of the entities:

```
@Entity({abstract: true})
export abstract class BaseUUIDEntity {

  @PrimaryKey({ type: 'uuid', defaultRaw: 'gen_random_uuid()' })
  uuid!: string;
}

@Entity({tableName: "tracks"})
export class Track extends BaseUUIDEntity {

    @OneToOne({entity: () => TrackMetric, inversedBy: 'track', nullable: true }) 
    metrics!: TrackMetric;
    ...
}

@Entity({tableName: "track_metrics"})
export class TrackMetric {

    @OneToOne({primary: true, fieldName: 'track_uuid', entity: () => Track, mappedBy: 'metrics', nullable: false})
    track!: Track;
    ...
}
```

The code I am using to persist looks as follows:

```
await RequestContext.createAsync(entity_manager, async () => {

       let track = await RequestContext.getEntityManager()!.findOne(Track, <uuid>);
       ...
       let metrics = new TrackMetric();
       track.metrics = metrics;
       metrics.track = track;
        
       await RequestContext.getEntityManager()!.persistAndFlush(track);
});
```

The error I am receiving is (I X'ed-out irrelevant data):
```
NotNullConstraintViolationException: insert into "track_metrics" ("xxx", "xxx", "xxx", "xxx") values (x, x, x, x) returning "track_uuid" - null value in column "track_uuid" of relation "track_metrics" violates not-null constraint
```

I was expecting the `track_uuid` field to be taken from the existing `uuid` in the `track` instance. Any suggestions as to what I am missing? I shouldn't have to explicitly set the `track_uuid` field myself, or do I? The documentation wasn't clear.

**Comment by B4nan** [comment, marked as answer]
Your entity definition seems wrong, you cant have a PK that is an inverse 1:1 relation - such relation is virtual, it does not represent any database column (unlike the owning side) - and PK needs to represent one.

> CREATE TABLE public.track_metrics (
>    track_uuid uuid NOT NULL,   // <-- This points to the 'owning' entity
>    ...

I think you got it wrong here, the side where you define the FK is called owning. And there is only one FK stored in 1:1 relation (on the owning side), not on boths ends. 

Looking at your schema, seems like you should just swap the `inversedBy` and `mappedBy`. 

```ts
@Entity()
export class Track {
  @PrimaryKey({ type: 'uuid', defaultRaw: 'gen_random_uuid()' })
  uuid!: string;

  // using `mappedBy` here - as this is the inverse side, there won't by any database column for this property
  @OneToOne({ entity: () => TrackMetric, mappedBy: 'track', nullable: true })
  metrics!: TrackMetric;
}

@Entity()
export class TrackMetric {
  // owning side, no need to use `inversedBy` here, it's enough to point here from the Track entity
  @OneToOne({ primary: true, entity: () => Track  })
  track!: Track;
}
```

You can always check what the schema generator produces for your entity definition (I would ideally suggest you use it exclusively instead of managing the schema manually - this way you ensure the entity definition is aligned and correct).
